drivers/adc/ad4170/ad4170.c

```diff
@@ -0,0 +1,2123 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include "no_os_error.h"
+#include "no_os_delay.h"
+#include "no_os_crc8.h"
+#include "ad4170.h"
+
+NO_OS_DECLARE_CRC8_TABLE(ad4170_crc8);
+
+struct ad4170_config ad4170_config_reset = {
+	.pin_muxing = {
+		.sync_ctrl = AD4170_SYNC_STANDARD,
+	},
+	.channel_en = 0x1,
+	.map = {
+		{.ainm = AD4170_AIN1},
+		{.ainm = AD4170_AIN1},
+		{.ainm = AD4170_AIN1},
+		{.ainm = AD4170_AIN1},
+		{.ainm = AD4170_AIN1},
+		{.ainm = AD4170_AIN1},
+		{.ainm = AD4170_AIN1},
+		{.ainm = AD4170_AIN1},
+		{.ainm = AD4170_AIN1},
+		{.ainm = AD4170_AIN1},
+		{.ainm = AD4170_AIN1},
+		{.ainm = AD4170_AIN1},
+		{.ainm = AD4170_AIN1},
+		{.ainm = AD4170_AIN1},
+		{.ainm = AD4170_AIN1},
+		{.ainm = AD4170_AIN1}
+	},
+	.setups = {
+		{ /* setup 0 */
+			.afe = {
+				.bipolar = true,
+				.ref_select = AD4170_REFIN_REFOUT,
+			},
+			.filter_fs = 0x4,
+			.gain = 0x555555,
+		},
+		{ /* setup 1 */
+			.afe = {
+				.bipolar = true,
+				.ref_select = AD4170_REFIN_REFOUT,
+			},
+			.filter_fs = 0x4,
+			.gain = 0x555555,
+		},
+		{ /* setup 2 */
+			.afe = {
+				.bipolar = true,
+				.ref_select = AD4170_REFIN_REFOUT,
+			},
+			.filter_fs = 0x4,
+			.gain = 0x555555,
+		},
+		{ /* setup 3 */
+			.afe = {
+				.bipolar = true,
+				.ref_select = AD4170_REFIN_REFOUT,
+			},
+			.filter_fs = 0x4,
+			.gain = 0x555555,
+		},
+		{ /* setup 4 */
+			.afe = {
+				.bipolar = true,
+				.ref_select = AD4170_REFIN_REFOUT,
+			},
+			.filter_fs = 0x4,
+			.gain = 0x555555,
+		},
+		{ /* setup 5 */
+			.afe = {
+				.bipolar = true,
+				.ref_select = AD4170_REFIN_REFOUT,
+			},
+			.filter_fs = 0x4,
+			.gain = 0x555555,
+		},
+		{ /* setup 6 */
+			.afe = {
+				.bipolar = true,
+				.ref_select = AD4170_REFIN_REFOUT,
+			},
+			.filter_fs = 0x4,
+			.gain = 0x555555,
+		},
+		{ /* setup 7 */
+			.afe = {
+				.bipolar = true,
+				.ref_select = AD4170_REFIN_REFOUT,
+			},
+			.filter_fs = 0x4,
+			.gain = 0x555555,
+		},
+	},
+	.ref_control = {
+		.ref_en = true,
+	},
+	.fir_control = {
+		.fir_length = 1,
+	}
+};
+
+static inline bool _is_addr_valid(struct ad4170_dev *dev, uint32_t reg_addr)
+{
+	uint8_t reg_size = AD4170_TRANSF_LEN(reg_addr);
+	uint32_t addr = AD4170_ADDR(reg_addr);
+
+	/* max readable address in short instruction mode */
+	if (dev->spi_settings.short_instruction && addr > 0x3F)
+		return false;
+
+	/* max valid address */
+#ifndef AD4170_TEST_AREA
+	if (addr > 0x195)
+		return false;
+#endif
+
+	if (reg_size > 4 || reg_size == 0)
+		return false;
+
+	return true;
+}
+
+static inline bool _is_big_endian()
+{
+	uint16_t a = 0x0100;
+	return (bool) *(uint8_t *) &a;
+}
+
+/* Swap bytes in a buffer with a given step
+ * Swap with step of 2:
+ * 	AA BB CC DD EE FF 00 11 becomes
+ *	BB AA DD CC FF EE 11 00
+ * Swap with step of 3:
+ * 	AA BB CC DD EE FF 00 11 22 becomes
+ * 	CC BB AA FF EE DD 22 11 00
+ * etc.
+ */
+static void _memswap64(void * buf, uint32_t bytes, size_t step)
```

Comment:
In the meantime, this function and `_is_bit_endian()` has been added to `no_os_util.c`, so we should use those instead.